<!DOCTYPE html>
<html>
<!-- Created by GNU Texinfo 7.0.1, https://www.gnu.org/software/texinfo/ -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<!-- This is Edition 1.6 of TCP/IP Internetworking with gawk,
for the 5.2.0 (or later) version of the GNU
implementation of AWK.


Copyright (C) 2000, 2001, 2002, 2004, 2009, 2010, 2016, 2019, 2020, 2021, 2023
Free Software Foundation, Inc.


Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.3 or
any later version published by the Free Software Foundation; with the
Invariant Sections being "GNU General Public License", the Front-Cover
texts being (a) (see below), and with the Back-Cover Texts being (b)
(see below).  A copy of the license is included in the section entitled
"GNU Free Documentation License".

a. "A GNU Manual"

b. "You have the freedom to
copy and modify this GNU manual.  Buying copies from the FSF
supports it in developing GNU and promoting software freedom." -->
<title>PROTBASE (TCP/IP Internetworking With gawk)</title>

<meta name="description" content="PROTBASE (TCP/IP Internetworking With gawk)">
<meta name="keywords" content="PROTBASE (TCP/IP Internetworking With gawk)">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta name="viewport" content="width=device-width,initial-scale=1">

<link href="index.html" rel="start" title="Top">
<link href="Index.html" rel="index" title="Index">
<link href="index.html#SEC_Contents" rel="contents" title="Table of Contents">
<link href="Some-Applications-and-Techniques.html" rel="up" title="Some Applications and Techniques">
<link href="STOXPRED.html" rel="prev" title="STOXPRED">
<style type="text/css">
<!--
div.center {text-align:center}
div.example {margin-left: 3.2em}
ul.mark-bullet {list-style-type: disc}
-->
</style>
<link rel="stylesheet" type="text/css" href="../../../../gnulib/manual.css">


</head>

<body lang="en">
<div class="section-level-extent" id="PROTBASE">
<div class="nav-panel">
<p>
Previous: <a href="STOXPRED.html" accesskey="p" rel="prev">STOXPRED: Stock Market Prediction As A Service</a>, Up: <a href="Some-Applications-and-Techniques.html" accesskey="u" rel="up">Some Applications and Techniques</a> &nbsp; [<a href="index.html#SEC_Contents" title="Table of contents" rel="contents">Contents</a>][<a href="Index.html" title="Index" rel="index">Index</a>]</p>
</div>
<hr>
<h3 class="section" id="PROTBASE_003a-Searching-Through-A-Protein-Database">3.10 PROTBASE: Searching Through A Protein Database</h3>
<a class="index-entry-id" id="index-PROTBASE"></a>
<a class="index-entry-id" id="index-NCBI_002c-National-Center-for-Biotechnology-Information"></a>
<a class="index-entry-id" id="index-BLAST_002c-Basic-Local-Alignment-Search-Tool"></a>
<a class="index-entry-id" id="index-Hoare_002c-C_002eA_002eR_002e-1"></a>
<blockquote class="quotation">
<p><i class="i">Inside every large problem is a small
problem struggling to get out.</i><a class="footnote" id="DOCF15" href="PROTBASE.html#FOOT15"><sup>15</sup></a>
</p></blockquote>
<div class="center">&mdash; <em class="emph">With apologies to C.A.R. Hoare</em>
</div>
<p>Yahoo&rsquo;s database of stock market data is just one among the many large
databases on the Internet. Another one is located at NCBI
(National Center for Biotechnology
Information). Established in 1988 as a national resource for molecular
biology information, NCBI creates public databases, conducts research
in computational biology, develops software tools for analyzing genome
data, and disseminates biomedical information. In this section, we
look at one of NCBI&rsquo;s public services, which is called BLAST
(Basic Local Alignment Search Tool).
</p>
<p>You probably know that the information necessary for reproducing living
cells is encoded in the genetic material of the cells. The genetic material
is a very long chain of four base nucleotides. It is the order of
appearance (the sequence) of nucleotides which contains the information
about the substance to be produced. Scientists in biotechnology often
find a specific fragment, determine the nucleotide sequence, and need
to know where the sequence at hand comes from.
</p>
<p>This is where the large
databases enter the game. At NCBI, databases store the knowledge
about which sequences have ever been found and where they have been found.
When the scientist sends his sequence to the BLAST service, the server
looks for regions of genetic material in its database which
look the most similar to the delivered nucleotide sequence. After a
search time of some seconds or minutes the server sends an answer to
the scientist. In order to make access simple, NCBI chose to offer
their database service through popular Internet protocols. There are
four basic ways to use the so-called BLAST services:
</p>
<ul class="itemize mark-bullet">
<li>The easiest way to use BLAST is through the web. Users may simply point
their browsers at the NCBI home page
and link to the BLAST pages.
NCBI provides a stable URL that may be used to perform BLAST searches
without interactive use of a web browser. This is what we will do later
in this section.
A demonstration client
and a <samp class="file">README</samp> file demonstrate how to access this URL.

</li><li>Currently,
<code class="command">blastcl3</code> is the standard network BLAST client.
You can download <code class="command">blastcl3</code> from the
anonymous FTP location.

</li><li>BLAST 2.0 can be run locally as a full executable and can be used to run
BLAST searches against private local databases, or downloaded copies of the
NCBI databases. BLAST 2.0 executables may be found on the NCBI
anonymous FTP server.

</li><li>The NCBI BLAST Email server is the best option for people without convenient
access to the web. A similarity search can be performed by sending a properly
formatted mail message containing the nucleotide or protein query sequence to
<a class="email" href="mailto:blast@ncbi.nlm.nih.gov">blast@ncbi.nlm.nih.gov</a>. The query sequence is compared against the
specified database using the BLAST algorithm and the results are returned in
an email message. For more information on formulating email BLAST searches,
you can send a message consisting of the word &ldquo;HELP&rdquo; to the same address,
<a class="email" href="mailto:blast@ncbi.nlm.nih.gov">blast@ncbi.nlm.nih.gov</a>.
</li></ul>

<p>Our starting point is the demonstration client mentioned in the first option.
The <samp class="file">README</samp> file that comes along with the client explains the whole
process in a nutshell. In the rest of this section, we first show
what such requests look like. Then we show how to use <code class="command">gawk</code> to
implement a client in about 10 lines of code. Finally, we show how to
interpret the result returned from the service.
</p>
<p>Sequences are expected to be represented in the standard
IUB/IUPAC amino acid and nucleic acid codes,
with these exceptions:  lower-case letters are accepted and are mapped
into upper-case; a single hyphen or dash can be used to represent a gap
of indeterminate length; and in amino acid sequences, &lsquo;<samp class="samp">U</samp>&rsquo; and &lsquo;<samp class="samp">*</samp>&rsquo;
are acceptable letters (see below).  Before submitting a request, any numerical
digits in the query sequence should either be removed or replaced by
appropriate letter codes (e.g., &lsquo;<samp class="samp">N</samp>&rsquo; for unknown nucleic acid residue
or &lsquo;<samp class="samp">X</samp>&rsquo; for unknown amino acid residue).
The nucleic acid codes supported are:
</p>
<div class="example">
<pre class="example-preformatted">A --&gt; adenosine               M --&gt; A C (amino)
C --&gt; cytidine                S --&gt; G C (strong)
G --&gt; guanine                 W --&gt; A T (weak)
T --&gt; thymidine               B --&gt; G T C
U --&gt; uridine                 D --&gt; G A T
R --&gt; G A (purine)            H --&gt; A C T
Y --&gt; T C (pyrimidine)        V --&gt; G C A
K --&gt; G T (keto)              N --&gt; A G C T (any)
                              -  gap of indeterminate length
</pre></div>

<p>Now you know the alphabet of nucleotide sequences. The last two lines
of the following example query show such a sequence, which is obviously
made up only of elements of the alphabet just described. Store this example
query into a file named <samp class="file">protbase.request</samp>. You are now ready to send
it to the server with the demonstration client.
</p>
<div class="example">
<pre class="example-preformatted">PROGRAM blastn
DATALIB month
EXPECT 0.75
BEGIN
&gt;GAWK310 the gawking gene GNU AWK
tgcttggctgaggagccataggacgagagcttcctggtgaagtgtgtttcttgaaatcat
caccaccatggacagcaaa
</pre></div>

<a class="index-entry-id" id="index-FASTA_002fPearson-format"></a>
<p>The actual search request begins with the mandatory parameter &lsquo;<samp class="samp">PROGRAM</samp>&rsquo;
in the first column followed by the value &lsquo;<samp class="samp">blastn</samp>&rsquo; (the name of the
program) for searching nucleic acids.  The next line contains the mandatory
search parameter &lsquo;<samp class="samp">DATALIB</samp>&rsquo; with the value &lsquo;<samp class="samp">month</samp>&rsquo; for the newest
nucleic acid sequences.  The third line contains an optional &lsquo;<samp class="samp">EXPECT</samp>&rsquo;
parameter and the value desired for it. The fourth line contains the
mandatory &lsquo;<samp class="samp">BEGIN</samp>&rsquo; directive, followed by the query sequence in
FASTA/Pearson format.
Each line of information must be less than 80 characters in length.
</p>
<p>The &ldquo;month&rdquo; database contains all new or revised sequences released in the
last 30 days and is useful for searching against new sequences.
There are five different blast programs, <code class="command">blastn</code> being the one that
compares a nucleotide  query  sequence  against a nucleotide sequence database.
</p>
<p>The last server directive that must appear in every request is the
&lsquo;<samp class="samp">BEGIN</samp>&rsquo; directive. The query sequence should immediately follow the
&lsquo;<samp class="samp">BEGIN</samp>&rsquo; directive and must appear in FASTA/Pearson format.
A sequence in
FASTA/Pearson format begins with a single-line description.
The description line, which is required, is distinguished from the lines of
sequence data that follow it by having a greater-than (&lsquo;<samp class="samp">&gt;</samp>&rsquo;) symbol
in the first column.  For the purposes of the BLAST server, the text of
the description is arbitrary.
</p>
<p>If you prefer to use a client written in <code class="command">gawk</code>, just store the following
10 lines of code into a file named <samp class="file">protbase.awk</samp> and use this client
instead. Invoke it with &lsquo;<samp class="samp">gawk -f protbase.awk protbase.request</samp>&rsquo;.
Then wait a minute and watch the result coming in. In order to replicate
the demonstration client&rsquo;s behavior as closely as possible, this client
does not use a proxy server. We could also have extended the client program
in <a class="ref" href="GETURL.html">Retrieving Web Pages</a>, to implement the client request from
<samp class="file">protbase.awk</samp> as a special case.
</p>
<div class="example smallexample">
<pre class="example-preformatted">{ request = request &quot;\n&quot; $0 }

END {
  BLASTService     = &quot;/inet/tcp/0/www.ncbi.nlm.nih.gov/80&quot;
  printf &quot;POST /cgi-bin/BLAST/nph-blast_report HTTP/1.0\n&quot; |&amp; BLASTService
  printf &quot;Content-Length: &quot; length(request) &quot;\n\n&quot;         |&amp; BLASTService
  printf request                                           |&amp; BLASTService
  while ((BLASTService |&amp; getline) &gt; 0)
      print $0
  close(BLASTService)
}
</pre></div>

<p>The demonstration client from NCBI is 214 lines long (written in C) and
it is not immediately obvious what it does. Our client is so short that
it <em class="emph">is</em> obvious what it does. First it loops over all lines of the
query and stores the whole query into a variable. Then the script
establishes an Internet connection to the NCBI server and transmits the
query by framing it with a proper HTTP request. Finally it receives
and prints the complete result coming from the server.
</p>
<p>Now, let us look at the result. It begins with an HTTP header, which you
can ignore. Then there are some comments about the query having been
filtered to avoid spuriously high scores. After this, there is a reference
to the paper that describes the software being used for searching the data
base. After a repetition of the original query&rsquo;s description we find the
list of significant alignments:
</p>
<div class="example smallexample">
<pre class="example-preformatted">Sequences producing significant alignments:                        (bits)  Value

gb|AC021182.14|AC021182 Homo sapiens chromosome 7 clone RP11-733...    38  0.20
gb|AC021056.12|AC021056 Homo sapiens chromosome 3 clone RP11-115...    38  0.20
emb|AL160278.10|AL160278 Homo sapiens chromosome 9 clone RP11-57...    38  0.20
emb|AL391139.11|AL391139 Homo sapiens chromosome X clone RP11-35...    38  0.20
emb|AL365192.6|AL365192 Homo sapiens chromosome 6 clone RP3-421H...    38  0.20
emb|AL138812.9|AL138812 Homo sapiens chromosome 11 clone RP1-276...    38  0.20
gb|AC073881.3|AC073881 Homo sapiens chromosome 15 clone CTD-2169...    38  0.20
</pre></div>

<p>This means that the query sequence was found in seven human chromosomes.
But the value 0.20 (20%) means that the probability of an accidental match
is rather high (20%) in all cases and should be taken into account.
You may wonder what the first column means. It is a key to the specific
database in which this occurrence was found.  The unique sequence identifiers
reported in the search results can be used as sequence retrieval keys
via the NCBI server. The syntax of sequence header lines used by the NCBI
BLAST server depends on the database from which each sequence was obtained.
The table below lists the identifiers for the databases from which the
sequences were derived.
</p>

<table class="multitable">
<tbody><tr><td>GenBank</td><td><code class="code">gb|accession|locus</code></td></tr>
<tr><td>EMBL Data Library</td><td><code class="code">emb|accession|locus</code></td></tr>
<tr><td>DDBJ, DNA Database of Japan</td><td><code class="code">dbj|accession|locus</code></td></tr>
<tr><td>NBRF PIR</td><td><code class="code">pir||entry</code></td></tr>
<tr><td>Protein Research Foundation</td><td><code class="code">prf||name</code></td></tr>
<tr><td>SWISS-PROT</td><td><code class="code">sp|accession|entry&nbsp;name<!-- /@w --></code></td></tr>
<tr><td>Brookhaven Protein Data Bank</td><td><code class="code">pdb|entry|chain</code></td></tr>
<tr><td>Kabat&rsquo;s Sequences of Immuno&hellip;</td><td><code class="code">gnl|kabat|identifier</code></td></tr>
<tr><td>Patents</td><td><code class="code">pat|country|number</code></td></tr>
<tr><td>GenInfo Backbone Id</td><td><code class="code">bbs|number</code></td></tr>
</tbody>
</table>


<p>For example, an identifier might be &lsquo;<samp class="samp">gb|AC021182.14|AC021182</samp>&rsquo;, where the
&lsquo;<samp class="samp">gb</samp>&rsquo; tag indicates that the identifier refers to a GenBank sequence,
&lsquo;<samp class="samp">AC021182.14</samp>&rsquo; is its GenBank ACCESSION, and &lsquo;<samp class="samp">AC021182</samp>&rsquo; is the GenBank LOCUS.
The identifier contains no spaces, so that a space indicates the end of the
identifier.
</p>
<p>Let us continue in the result listing. Each of the seven alignments mentioned
above is subsequently described in detail. We will have a closer look at
the first of them.
</p>
<div class="example smallexample">
<pre class="example-preformatted">&gt;gb|AC021182.14|AC021182 Homo sapiens chromosome 7 clone RP11-733N23, WORKING DRAFT SEQUENCE, 4
             unordered pieces
          Length = 176383

 Score = 38.2 bits (19), Expect = 0.20
 Identities = 19/19 (100%)
 Strand = Plus / Plus

Query: 35    tggtgaagtgtgtttcttg 53
             |||||||||||||||||||
Sbjct: 69786 tggtgaagtgtgtttcttg 69804
</pre></div>

<p>This alignment was located on the human chromosome 7. The fragment on which
part of the query was found had a total length of 176383. Only 19 of the
nucleotides matched and the matching sequence ran from character 35 to 53
in the query sequence and from 69786 to 69804 in the fragment on chromosome 7.
If you are still reading at this point, you are probably interested in finding
out more about Computational Biology and you might appreciate the following
hints.
</p>
<a class="index-entry-id" id="index-Computational-Biology"></a>
<a class="index-entry-id" id="index-Bioinformatics"></a>
<ol class="enumerate">
<li> There is a book called <cite class="cite">Introduction to Computational Biology</cite>
by Michael S. Waterman, which is worth reading if you are seriously
interested. You can find a good
book review
on the Internet.

</li><li> While Waterman&rsquo;s book explains the algorithms employed internally
in the database search engines, most practitioners prefer to approach
the subject differently. The applied side of Computational Biology is
called Bioinformatics, and emphasizes the tools available for day-to-day
work as well as how to actually <em class="emph">use</em> them. One of the very few affordable
books on Bioinformatics is
<cite class="cite">Developing Bioinformatics Computer Skills</cite>.

</li><li> The sequences <em class="emph">gawk</em> and <em class="emph">gnuawk</em> are in widespread use in
the genetic material of virtually every earthly living being. Let us
take this as a clear indication that the divine creator has intended
<code class="command">gawk</code> to prevail over other scripting languages such as &lsquo;<samp class="samp">perl</samp>&rsquo;,
&lsquo;<samp class="samp">tcl</samp>&rsquo;, or &lsquo;<samp class="samp">python</samp>&rsquo; which are not even proper sequences. (:-)
</li></ol>

</div>
<div class="footnotes-segment">
<hr>
<h4 class="footnotes-heading">Footnotes</h4>

<h5 class="footnote-body-heading"><a id="FOOT15" href="PROTBASE.html#DOCF15">(15)</a></h5>
<p>What C.A.R. Hoare
actually said was &ldquo;Inside every large program is a
small program struggling to get out.&rdquo;</p>
</div>
<hr>
<div class="nav-panel">
<p>
Previous: <a href="STOXPRED.html">STOXPRED: Stock Market Prediction As A Service</a>, Up: <a href="Some-Applications-and-Techniques.html">Some Applications and Techniques</a> &nbsp; [<a href="index.html#SEC_Contents" title="Table of contents" rel="contents">Contents</a>][<a href="Index.html" title="Index" rel="index">Index</a>]</p>
</div>



</body>
</html>
